### ISSUE TITLE
Lazily-evaluated relationships

### ISSUE THREAD

**Issue by mwpastore** [fotinakis/jsonapi-serializers#29]
I'm performance tuning some of my endpoints and I've noticed that JSONAPI::Serializers is triggering database calls on relationships even when those relationships are not given in the `included` option (i.e. `included` is `nil`). The results of these database calls are not being included in the output. Is this intentional, and if so, is there a way to tell JSONAPI::Serializers to lazily-evaluate relationships? If not intentional, is it possible there is some some class instantiation logic that is inspecting or otherwise poking the relationships that is causing the upstream Sequel::Model to fire off queries?

```ruby
require 'sequel'
require 'jsonapi-serializers'
require 'logger'
require 'pp'

DB = Sequel.sqlite nil, 
  :loggers => [Logger.new($stdout)]

DB.create_table :foos do
  primary_key :id
  foreign_key :bar_id, :bars
  String :name
end

DB.create_table :bars do
  primary_key :id
end

DB.run "INSERT INTO bars (id) VALUES (1)"
DB.run "INSERT INTO foos (id, bar_id, name) VALUES (1, 1, 'myfoo')"

class Foo < Sequel::Model
  many_to_one :bar
end

class Bar < Sequel::Model
  one_to_many :foos
end

class FooSerializer
  include JSONAPI::Serializer

  attribute :name

  has_one :bar
end

class BarSerializer
  include JSONAPI::Serializer

  has_many :foos
end

pp JSONAPI::Serializer.serialize(Foo.first, {
  :is_collection => false,
  :skip_collection_check => true
})
```

And the output:

```
$ bundle show jsonapi-serializers
/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0
$ bundle show sequel
/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0
$ ruby --version
ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-linux]
$ ruby main.rb 
I, [2015-10-13T15:01:17.021499 #24657]  INFO -- : (0.000207s) PRAGMA foreign_keys = 1
I, [2015-10-13T15:01:17.021745 #24657]  INFO -- : (0.000031s) PRAGMA case_sensitive_like = 1
I, [2015-10-13T15:01:17.022125 #24657]  INFO -- : (0.000270s) CREATE TABLE `foos` (`id` integer NOT NULL PRIMARY KEY AUTOINCREMENT, `bar_id` integer REFERENCES `bars`, `name` varchar(255))
I, [2015-10-13T15:01:17.022445 #24657]  INFO -- : (0.000118s) CREATE TABLE `bars` (`id` integer NOT NULL PRIMARY KEY AUTOINCREMENT)
I, [2015-10-13T15:01:17.022600 #24657]  INFO -- : (0.000053s) INSERT INTO bars (id) VALUES (1)
I, [2015-10-13T15:01:17.022759 #24657]  INFO -- : (0.000053s) INSERT INTO foos (id, bar_id, name) VALUES (1, 1, 'myfoo')
I, [2015-10-13T15:01:17.023645 #24657]  INFO -- : (0.000307s) PRAGMA table_info('foos')
I, [2015-10-13T15:01:17.024828 #24657]  INFO -- : (0.000089s) PRAGMA table_info('bars')
I, [2015-10-13T15:01:17.025724 #24657]  INFO -- : (0.000076s) SELECT sqlite_version()
I, [2015-10-13T15:01:17.026072 #24657]  INFO -- : (0.000168s) SELECT * FROM `foos` LIMIT 1
I, [2015-10-13T15:01:17.026730 #24657]  INFO -- : (0.000078s) SELECT * FROM `bars` WHERE `id` = 1
{"data"=>
  {"id"=>"1",
   "type"=>"foos",
   "attributes"=>{"name"=>"myfoo"},
   "links"=>{"self"=>"/foos/1"},
   "relationships"=>
    {"bar"=>
      {"links"=>
        {"self"=>"/foos/1/relationships/bar", "related"=>"/foos/1/bar"}}}}}
```

The last INFO line (```SELECT * FROM `bars` WHERE `id` = 1```) is what I'm trying to eliminate.

**Comment by mwpastore** [comment]
For the sake of completeness, here it is with the bar included. Note the extraneous query to go back and get the foos for the bar, even though I did not list `bar.foos` in `included`.

```
$ ruby main.rb 
I, [2015-10-13T15:11:27.749979 #25028]  INFO -- : (0.000212s) PRAGMA foreign_keys = 1
I, [2015-10-13T15:11:27.750209 #25028]  INFO -- : (0.000030s) PRAGMA case_sensitive_like = 1
I, [2015-10-13T15:11:27.750652 #25028]  INFO -- : (0.000322s) CREATE TABLE `foos` (`id` integer NOT NULL PRIMARY KEY AUTOINCREMENT, `bar_id` integer REFERENCES `bars`, `name` varchar(255))
I, [2015-10-13T15:11:27.750986 #25028]  INFO -- : (0.000136s) CREATE TABLE `bars` (`id` integer NOT NULL PRIMARY KEY AUTOINCREMENT)
I, [2015-10-13T15:11:27.751187 #25028]  INFO -- : (0.000086s) INSERT INTO bars (id) VALUES (1)
I, [2015-10-13T15:11:27.751377 #25028]  INFO -- : (0.000056s) INSERT INTO foos (id, bar_id, name) VALUES (1, 1, 'myfoo')
I, [2015-10-13T15:11:27.752178 #25028]  INFO -- : (0.000192s) PRAGMA table_info('foos')
I, [2015-10-13T15:11:27.753282 #25028]  INFO -- : (0.000081s) PRAGMA table_info('bars')
I, [2015-10-13T15:11:27.754148 #25028]  INFO -- : (0.000063s) SELECT sqlite_version()
I, [2015-10-13T15:11:27.754462 #25028]  INFO -- : (0.000144s) SELECT * FROM `foos` LIMIT 1
I, [2015-10-13T15:11:27.755101 #25028]  INFO -- : (0.000078s) SELECT * FROM `bars` WHERE `id` = 1
I, [2015-10-13T15:11:27.757103 #25028]  INFO -- : (0.000101s) SELECT * FROM `foos` WHERE (`foos`.`bar_id` = 1)
{"data"=>
  {"id"=>"1",
   "type"=>"foos",
   "attributes"=>{"name"=>"myfoo"},
   "links"=>{"self"=>"/foos/1"},
   "relationships"=>
    {"bar"=>
      {"links"=>
        {"self"=>"/foos/1/relationships/bar", "related"=>"/foos/1/bar"},
       "data"=>{"type"=>"bars", "id"=>"1"}}}},
 "included"=>
  [{"id"=>"1",
    "type"=>"bars",
    "attributes"=>{},
    "links"=>{"self"=>"/bars/1"},
    "relationships"=>
     {"foos"=>
       {"links"=>
         {"self"=>"/bars/1/relationships/foos", "related"=>"/bars/1/foos"}}}}]}
```

**Comment by fotinakis** [comment]
Consolidating this issue into https://github.com/fotinakis/jsonapi-serializers/issues/30

**Comment by mwpastore** [comment]
Mike, can you help me understand how these two issues are related? Why does JSONAPI::Serializers need to know `bar.id` (in this example) if `bar` isn't being included?

**Comment by fotinakis** [comment]
@mwpastore -- you're right, I'm not sure where we're loading this one. :| I thought it related to how I implemented the requirements around resource linkage (See http://jsonapi.org/format/#document-resource-object-linkage and the distinction between empty/non-empty relationships) ... but, looking deeper here there is no resource linkage, only the `links` data which should be able to be computed without the object lookup.

I've looked through the code for a minute but can't see exactly where we touch the object. Any chance you can drop some debugging into your test code and find where in jsonapi-serializers we touch the related object?

**Comment by mwpastore** [comment]
Here's what I was able to coax out of Sequel: 

```
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/database/logging.rb:42:in `log_yield'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/adapters/sqlite.rb:185:in `block in _execute'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/database/connecting.rb:249:in `block in synchronize'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/connection_pool/threaded.rb:103:in `hold'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/database/connecting.rb:249:in `synchronize'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/adapters/sqlite.rb:178:in `_execute'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/adapters/sqlite.rb:128:in `execute'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/dataset/actions.rb:921:in `execute'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/adapters/sqlite.rb:319:in `fetch_rows'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/model/base.rb:1043:in `primary_key_lookup'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/model/associations.rb:2124:in `_load_associated_object_via_primary_key'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/model/associations.rb:2144:in `_load_associated_objects'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/model/associations.rb:2240:in `load_associated_objects'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/sequel-4.25.0/lib/sequel/model/associations.rb:1767:in `block in def_association_method'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:198:in `evaluate_attr_or_block'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:168:in `block in has_one_relationships'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:166:in `each'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:166:in `has_one_relationships'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:94:in `relationships'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:323:in `serialize_primary'",
 "/home/mwp/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/jsonapi-serializers-0.3.0/lib/jsonapi-serializers/serializer.rb:272:in `serialize'"
```

**Comment by dominic** [comment]
Just experience the same issue while attempting a patch for improved `context` passing. The culprit seems to be here: https://github.com/fotinakis/jsonapi-serializers/blob/68c6dc227ee2b32dc484017a1ac13aecb1d8eab7/lib/jsonapi-serializers/serializer.rb#L192-L200

Big :+1: to https://github.com/fotinakis/jsonapi-serializers/pull/31

**Comment by mwpastore** [comment]
Fix merged per @fotinakis and included in v0.3.1.
